geomstats/geometry/pullback_metric.py

```diff
@@ -125,34 +174,94 @@ def inner_product_derivative_matrix(self, base_point=None):
 
-            def immersion_a(x):
-                return self.immersion(x)[a]
+            hessian_aij = self.hessian_immersion(point)
+            if self.dim == 1 and hessian_aij.ndim > 3:
+                hessian_aij = gs.squeeze(hessian_aij, axis=-1)
 
-            jacobian_a, hessian_a = gs.autodiff.jacobian_and_hessian(immersion_a)(
-                base_point
-            )
-            if self.dim == 1 and hessian_a.ndim > 2:
-                hessian_a = gs.squeeze(hessian_a, axis=-1)
-            hessian_aij.append(hessian_a)
-
-            jacobian_a = gs.squeeze(jacobian_a, axis=0)
-            if len(jacobian_a.shape) == 0:
-                jacobian_a = gs.to_ndarray(jacobian_a, to_ndim=1)
-            jacobian_ai.append(jacobian_a)
-
-        hessian_aij = gs.stack(hessian_aij, axis=0)
-        jacobian_ai = gs.stack(jacobian_ai, axis=0)
-        inner_prod_deriv_mat = gs.einsum(
-            "aki,aj->kij", hessian_aij, jacobian_ai
-        ) + gs.einsum("akj,ai->kij", hessian_aij, jacobian_ai)
-
-        inner_prod_deriv_mat = gs.transpose(inner_prod_deriv_mat, axes=(2, 1, 0))
-        return inner_prod_deriv_mat
+            inner_prod_deriv_mat = gs.einsum(
+                "aki,aj->ijk", hessian_aij, jacobian_ai
+            ) + gs.einsum("akj,ai->ijk", hessian_aij, jacobian_ai)
+            inner_prod_deriv_mats.append(inner_prod_deriv_mat)
+
+        inner_prod_deriv_mat = gs.stack(inner_prod_deriv_mats, axis=0)
+        return inner_prod_deriv_mat[0] if initial_ndim == 1 else inner_prod_deriv_mat
+
+    def second_fundamental_form(self, base_point):
+        r"""Compute the second fundamental form.
+
+        Parameters
+        ----------
+        base_point : array-like, shape=[..., dim]
+            Base point.
+
+        Returns
+        -------
+        second_fundamental_form : array-like, shape=[..., embedding_dim, dim, dim]
+            Second fundamental form.
+        """
+        initial_ndim = base_point.ndim
+        base_point = gs.to_ndarray(base_point, to_ndim=2)
+        second_fundamental_forms = []
+        for point in base_point:
+            christoffels = self.christoffels(point)
+
+            jacobian_ai = self.jacobian_immersion(point)
+            if self.dim == 1 and jacobian_ai.ndim > 2:
+                jacobian_ai = gs.squeeze(jacobian_ai, axis=-1)
+
+            hessian_aij = self.hessian_immersion(point)
+            if self.dim == 1 and hessian_aij.ndim > 3:
+                hessian_aij = gs.squeeze(hessian_aij, axis=-1)
+
+            second_fundamental_form_aij = []
+            for a in range(self.embedding_metric.dim):
+                jacobian_a = jacobian_ai[a]
+                hessian_a = hessian_aij[a]
+
+                second_fundamental_form_a = hessian_a - gs.einsum(
+                    "kij,k->ij", christoffels, jacobian_a
+                )
+                second_fundamental_form_aij.append(second_fundamental_form_a)
```

Comment:
We may be able to vectorize here, no?

Something like `second_fundamental_form_aij = hessian_aij - gs.einsum("...kij,...k->...ij", christoffels, jacobian_a)` should work.